### PR TITLE
STM32U0: provide support for I2C peripherals 

### PR DESCRIPTION
**Pull request by djiatsaf-st**
This PR adds I2C peripherals on the stm32U0 boards.

### CODE REVIEW

File: dts/arm/st/u0/stm32u0.dtsi [hunk 1]
```diff
@@ -207,6 +208,42 @@
...
+		};
+
+		i2c3: i2c@40008800 {
+			compatible = "st,stm32u0-i2c", "st,stm32-i2c-v2";
+			clock-frequency = <I2C_BITRATE_STANDARD>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40008800 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00800000>;
+			interrupts = <24 0>;
```
Comment (erwango) [inline dts/arm/st/u0/stm32u0.dtsi:243]: Question here.
It seems that instances 2, 3 and 4 will share the same interrupt line. Issue is that there is nothing to handle this case from what I can see at first sight.
Please confirm and see how to address:
- add support using SHARED_IRQ
- generate an error on unsupported configuration
- ...

File: boards/st/nucleo_u031r8/doc/index.rst
Comment (erwango) [inline boards/st/nucleo_u031r8/doc/index.rst:1]: please update board's .yaml file

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/i2c/st,stm32u0-i2c.yaml
Comment (erwango) [inline dts/bindings/i2c/st,stm32u0-i2c.yaml:1]: This commit should come first to allow compilation of others.

File: dts/bindings/i2c/st,stm32u0-i2c.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2024 STMicroelectronics
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+        STM32U0 I2C Controller
+        This compatibility applies to all I2C hardware blocks corresponding
+        to the version available in STM32U0 SoCs.
+        This version of the STM32 I2C hardware block can be identified by the
+        absence of the SMBHEN and SMBDEN bits in the I2C_CR1 register.
+        For example, but not limited to: STM32U0
```
Comment (erwango) [inline dts/bindings/i2c/st,stm32u0-i2c.yaml:10]: Could be summarized by "[...] doesn't support SMBUS"

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_ll_stm32.c [hunk 3]
```diff
@@ -115,7 +115,9 @@ int i2c_stm32_runtime_configure(const struct device *dev, uint32_t config)
 #endif
 
 	LL_I2C_Disable(i2c);
+#if !DT_HAS_COMPAT_STATUS_OKAY(st_stm32u0_i2c)
```
Comment (ydamigos) [inline drivers/i2c/i2c_ll_stm32.c:118]: Are we going to follow this approach or check if the specific bits are defined in SoC's header file (e.g. `I2C_CR1_SMBDEN`) like we do elsewhere in I2C driver (e.g. `I2C_FLTR_ANOFF`)?

**Review by ydamigos — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_ll_stm32.c (hunk 3 above)
Comment (djiatsaf-st) [reply drivers/i2c/i2c_ll_stm32.c:118]: Hello @ydamigos ,
Thanks for the suggestion, I think I'll do it this way, it works and I don't have any other constraints with it.

Comment (ydamigos) [reply drivers/i2c/i2c_ll_stm32.c:118]: I would avoid to introduce a new compatible `st,stm32u0-i2c`, just to exclude code, while there is another way doing it which is already there and it is used in STM32 I2C driver. But @erwango has the last word as the maintainer.

Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c:118]: Thanks @ydamigos for the proposal. I agree that for now, a new compatible doesn't add much value and is more complex. Let's go with the register presence check for now, we'll still be able to come back on that later if required. Thanks for the proposal, @ydamigos!

Comment (djiatsaf-st) [reply drivers/i2c/i2c_ll_stm32.c:118]: > I would avoid to introduce a new compatible `st,stm32u0-i2c`, just to exclude code, while there is another way doing it which is already there and it is used in STM32 I2C driver. But @erwango has the last word as the maintainer.

hello @ydamigos 
I think my last post was confusing, I said I tested your suggestion and it worked, so I would use that instead , it's simpler indeed.

File: soc/st/stm32/stm32u0x/Kconfig.defconfig.stm32u073xx [hunk 4]
```diff
@@ -8,4 +8,24 @@ if SOC_STM32U073XX
...
+
+config I2C3_I2C4_ENABLED
+	default y if $(dt_nodelabel_enabled,i2c3) && $(dt_nodelabel_enabled,i2c4)
+
+config I2C2_I2C3_I2C4_ENABLED
+	default y if $(dt_nodelabel_enabled,i2c2) && $(dt_nodelabel_enabled,i2c3) &&  $(dt_nodelabel_enabled,i2c4)
+
+
+config SHARED_INTERRUPTS
+	default y if I2C2_I2C3_ENABLED || I2C3_I2C4_ENABLED || I2C2_I2C4_ENABLED || I2C2_I2C3_I2C4_ENABLED
```
Comment (erwango) [inline soc/st/stm32/stm32u0x/Kconfig.defconfig.stm32u073xx:27]: Could be summed up as:
```suggestion
config SHARED_INTERRUPTS
	default y if $(dt_nodelabel_enabled,i2c2) && $(dt_nodelabel_enabled,i2c3)
	default y if $(dt_nodelabel_enabled,i2c2) && $(dt_nodelabel_enabled,i2c4)
	default y if $(dt_nodelabel_enabled,i2c3) && $(dt_nodelabel_enabled,i2c4)
```

and  can be put in `soc/st/stm32/stm32u0x/Kconfig.defconfig`

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_ll_stm32.c [hunk 5]
```diff
@@ -115,7 +115,9 @@ int i2c_stm32_runtime_configure(const struct device *dev, uint32_t config)
 #endif
 
 	LL_I2C_Disable(i2c);
+#if defined(SMBDEN) && defined(SMBHEN)
```
Comment (ydamigos) [inline drivers/i2c/i2c_ll_stm32.c:118]:
```suggestion
#if defined(I2C_CR1_SMBHEN) && defined(I2C_CR1_SMBHEN)
```

**Review by ydamigos — CHANGES_REQUESTED** [review]

**Review by ydamigos — APPROVED** [review]
LGTM

**Review by erwango — APPROVED** [review]

**Pull request merged**